File: app/src/main/java/com/jakebarnby/filemanager/sources/dropbox/DropboxFactory.java
package com.jakebarnby.filemanager.sources.dropbox;

import android.content.Context;
import android.os.AsyncTask;

import com.dropbox.core.DbxException;
import com.dropbox.core.v2.DbxClientV2;
import com.dropbox.core.v2.files.FileMetadata;
import com.dropbox.core.v2.files.FolderMetadata;
import com.dropbox.core.v2.files.Metadata;
import com.dropbox.core.v2.files.WriteMode;
import com.dropbox.core.v2.users.SpaceAllocation;
import com.dropbox.core.v2.users.SpaceUsage;
import com.google.firebase.analytics.FirebaseAnalytics;
import com.jakebarnby.filemanager.sources.models.SourceStorageStats;
import com.jakebarnby.filemanager.util.Constants;
import com.jakebarnby.filemanager.util.PreferenceUtils;
import com.jakebarnby.filemanager.util.Utils;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

/**
 * Created by Jake on 6/9/2017.
 */

public class DropboxFactory {

    private static final DropboxFactory ourInstance = new DropboxFactory();
    private static final String TAG = "DROPBOX";
    private DbxClientV2 mClient;

    public static DropboxFactory getInstance() {
        return ourInstance;
    }

    public DbxClientV2 getClient() {
        return mClient;
    }

    public void setClient(DbxClientV2 mClient) {
        this.mClient = mClient;
    }

    /**
     * @param downloadPath
     * @param destinationPath
     * @return
     */
    public File downloadFile(String downloadPath, String destinationPath) throws IOException, DbxException {
        File file = new File(destinationPath);
        OutputStream outputStream = new FileOutputStream(file);
        getClient()
                .files()
                .download(downloadPath)
                .download(outputStream);
        return file;
    }

    /**
     * @param fileUri
     * @param destPath
     */
    public FileMetadata uploadFile(String fileUri, String destPath) throws IOException, DbxException {
        File localFile = new File(fileUri);
        if (localFile.exists()) {
            InputStream inputStream = new FileInputStream(localFile);
            return getClient()
                    .files()
                    .uploadBuilder(destPath + "/" + localFile.getName())
                    .withMode(WriteMode.OVERWRITE)
                    .uploadAndFinish(inputStream);
        }
        return null;
    }

    /**
     * @param filePath
     */
    public void deleteFile(String filePath) throws DbxException {
        getClient()
                .files()
                .delete(filePath);
    }

    /**
     * @param name
     * @param path
     */
    public FolderMetadata createFolder(String name, String path) throws DbxException {
        return getClient()
                .files()
                .createFolder(path + File.separator + name);
    }

    /**
     * @param oldPath
     * @param newPath
     * @return
     */
    public Metadata rename(String oldPath, String newPath) throws DbxException {
        return getClient()
                .files()
                .move(oldPath, newPath);
    }

    public void logout(Context context) {
        AsyncTask.execute(() -> {
            try {
                getClient().auth().tokenRevoke();
                setClient(null);

                PreferenceUtils.savePref(context, Constants.Prefs.DROPBOX_TOKEN_KEY, (String) null);
            } catch (DbxException e) {
                Utils.logFirebaseErrorEvent(
                        FirebaseAnalytics.getInstance(context),
                        Constants.Analytics.EVENT_ERROR_DROPBOX_LOGOUT,
                        e.getMessage() == null ? e.toString() : e.getMessage());
            }
        });
    }

    public SourceStorageStats getStorageStats() {
        try {
            SpaceUsage usage = getClient().users().getSpaceUsage();
            long used = usage.getUsed();
            long max = 0;

            SpaceAllocation alloc = usage.getAllocation();
            if (alloc.isIndividual()) {
                max += alloc.getIndividualValue().getAllocated();
            }

            if (alloc.isTeam()) {
                max += alloc.getTeamValue().getAllocated();
            }

            SourceStorageStats info = new SourceStorageStats();
            info.setTotalSpace(max);
            info.setUsedSpace(used);
            info.setFreeSpace(max - used);

            return info;
        } catch (DbxException e) {
            e.printStackTrace();
        }
        return null;
    }
}

File: app/src/main/java/com/jakebarnby/filemanager/sources/dropbox/DropboxFactory.kt
package com.jakebarnby.filemanager.sources.dropbox

import android.content.Context
import androidx.core.os.bundleOf
import com.dropbox.core.DbxException
import com.dropbox.core.v2.DbxClientV2
import com.dropbox.core.v2.files.FileMetadata
import com.dropbox.core.v2.files.FolderMetadata
import com.dropbox.core.v2.files.Metadata
import com.dropbox.core.v2.files.WriteMode
import com.google.firebase.analytics.FirebaseAnalytics
import com.jakebarnby.filemanager.sources.models.SourceStorageStats
import com.jakebarnby.filemanager.util.Constants
import com.jakebarnby.filemanager.util.Constants.Prefs
import com.jakebarnby.filemanager.util.Logger
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException

/**
 * Created by Jake on 6/9/2017.
 */
object DropboxFactory {

    private const val TAG = "DROPBOX"

    var client: DbxClientV2? = null

    /**
     * @param downloadPath
     * @param destinationPath
     * @return
     */
    @Throws(IOException::class, DbxException::class)
    fun downloadFile(
        downloadPath: String,
        destinationPath: String
    ): File {
        val file = File(destinationPath)
        FileOutputStream(file).use {
            client?.files()
                ?.download(downloadPath)
                ?.download(it)
        }
        return file
    }

    /**
     * @param fileUri
     * @param destPath
     */
    @Throws(IOException::class, DbxException::class)
    fun uploadFile(
        fileUri: String,
        destPath: String
    ): FileMetadata? {
        val localFile = File(fileUri)
        if (!localFile.exists()) {
            return null
        }

        FileInputStream(localFile).use {
            return client?.files()
                ?.uploadBuilder(destPath + "/" + localFile.name)
                ?.withMode(WriteMode.OVERWRITE)
                ?.uploadAndFinish(it)
        }

    }

    /**
     * @param filePath
     */
    @Throws(DbxException::class)
    fun deleteFile(filePath: String?) {
        client?.files()
            ?.delete(filePath)
    }

    /**
     * @param name
     * @param path
     */
    @Throws(DbxException::class)
    fun createFolder(name: String?, path: String?): FolderMetadata? =
        client?.files()?.createFolder(path + File.separator + name)

    /**
     * @param oldPath
     * @param newPath
     * @return
     */
    @Throws(DbxException::class)
    fun rename(oldPath: String?, newPath: String?): Metadata? =
        client?.files()?.move(oldPath, newPath)

    fun logout(context: Context) {
        try {
            client!!.auth().tokenRevoke()
            client = null
            Preferences.savePref(context, Prefs.DROPBOX_TOKEN_KEY, null as String?)
        } catch (e: DbxException) {
            val params = bundleOf(Constants.Analytics.PARAM_ERROR_VALUE to e.message)
            Logger.logFirebaseEvent(
                FirebaseAnalytics.getInstance(context),
                Constants.Analytics.EVENT_ERROR_DROPBOX_LOGOUT,
                params
            )
        }
    }

    val storageStats: SourceStorageStats?
        get() {
            try {
                val usage = client!!.users().spaceUsage
                val used = usage.used
                var max: Long = 0
                val alloc = usage.allocation
                if (alloc.isIndividual) {
                    max += alloc.individualValue.allocated
                }
                if (alloc.isTeam) {
                    max += alloc.teamValue.allocated
                }
                val info = SourceStorageStats()
                info.totalSpace = max
                info.usedSpace = used
                info.freeSpace = max - used
                return info
            } catch (e: DbxException) {
                e.printStackTrace()
            }
            return null
        }
}